[vincesum] iP

vincesum opened the pull request.

vincesum committed:
Renamed chatbot to Vinbot

Added greeting and bye message

vincesum committed:
Add echo functionality

Added function to switch off bot when user inputs "bye". Added personality to the bot by adding emptyline.

vincesum committed:
Add addtask functionality to store task in array

Add list function to display current tasks
Updated the indentation and lines of the bot to fit personality

vincesum committed:
Add Task class

Changed array of String to array of Task to store tasks
Add Scanner to scan for "mark" and "unmark" to mark and unmark tasks as done
Add descriptive instructions for users to use the bot

vincesum committed:
Followed the coding standards

vincesum committed:
Fixed mark and unmark bug

Bug triggered when entering non-integer value after "mark" or "unmark"

vincesum committed:
Implemented SLAP into the algorithm by creating new class LineReader.java

Extracted methods like welcoming the user out so the code looks more readable

vincesum committed:
Added Todos, Deadlines, Events classes as subclasses of Task

Created toString method in Task
Use method overriding to override toString() in Todos
Add getLabel() to Task and all subclasses to return appropriate label
Todos is working properly

vincesum committed:
Fixed bug in Todos

vincesum committed:
Modify toString to include getDate method in order to print date for deadline

Add getDate() in Task and subclasses to print date for list method
Add deadline case in LineReader to scan for "deadline" and check if deadLineData received is of appropriate format. If not, print error message
Add deadLineData array to store deadline data of description and due date
Add restrictions for deadline input data to only accept deadlines that contains "/by"

vincesum committed:
Add printHelpMessage method for users to read the list of commands

Break down the string received for event and store the description, start and end dates into an array eventData.
Add event case in LineReader to scan for "event" and check if eventData received is of appropriate format. If not, print error message

vincesum committed:
Fix LineReader bug where it does not read more than 1 line from input.txt

Added some test cases into input to test functionality of Vinbot
Updated EXPECTED.TXT to determine correct output

vincesum committed:
Implemented SLAP for scanMessage by extracting methods, reducing number of lines from 60+ lines to 24

Followed coding standards by avoiding deep nesting
Paragraphing of code done
Change naming of result to format
Remove redundant toString method in Task and its subclasses
Made the happy path prominent

Basudeb2005 commented on src/main/java/Deadlines.java:
```diff
@@ -0,0 +1,32 @@
+            return null;
+        }
+
+        String desc = line.substring(0 , line.indexOf("/by"));
```
There is an unnecessary space before the comma in substring(0 , line.indexOf("/by"));. It should be substring(0, line.indexOf("/by")); to maintain proper Java coding conventions.

Basudeb2005 replied:
There is an unnecessary space before the comma in substring(0 , line.indexOf("/by"));. It should be substring(0, line.indexOf("/by")); to maintain proper Java coding conventions.

Basudeb2005 commented on src/main/java/Events.java:
```diff
@@ -0,0 +1,37 @@
+            return null;
+        }
+
+        String desc = line.substring(0 , line.indexOf("/from"));
```
Similar to Deadlines.java, there is an extra space before the comma. Use substring(0, line.indexOf("/from")); to follow proper formatting.

Basudeb2005 replied:
Similar to Deadlines.java, there is an extra space before the comma. Use substring(0, line.indexOf("/from")); to follow proper formatting.

Basudeb2005 commented on src/main/java/LineReader.java:
```diff
@@ -0,0 +1,122 @@
+        isActive = status;
+    }
+
+    public void scanMessage(boolean status, Vinbot.printWelcomeMessage format, Task[] storage) {
```
The method signature has a boolean parameter named status. Since it is also used in a loop to track the active state, consider renaming it to something more descriptive like isRunning to improve readability.

Basudeb2005 replied:
The method signature has a boolean parameter named status. Since it is also used in a loop to track the active state, consider renaming it to something more descriptive like isRunning to improve readability.

Basudeb2005 commented on src/main/java/LineReader.java:
```diff
@@ -0,0 +1,122 @@
+            return;
+        }
+
+        int integer = Integer.parseInt(intValue) - 1;
```
Consider using a more meaningful variable name instead of integer, such as taskIndex, to enhance clarity.

Basudeb2005 replied:
Consider using a more meaningful variable name instead of integer, such as taskIndex, to enhance clarity.

weiliangng commented on src/main/java/LineReader.java:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    private int handleMessage(Vinbot.printWelcomeMessage format, Task[] storage, String line, int numberOfElements) {
+        if (line.toLowerCase().startsWith("mark") || line.toLowerCase().startsWith("unmark")) {
+            handleMarkMessage(storage, line, numberOfElements);
+        }
+        else if (line.toLowerCase().startsWith("todo")) {
+            numberOfElements = handleTodoMessage(format, storage, line, numberOfElements);
+        }
+        else if (line.toLowerCase().startsWith("deadline")) {
+            numberOfElements = handleDeadLineMessage(format, storage, line, numberOfElements);
+        }
+        else if (line.toLowerCase().startsWith("event")) {
+            numberOfElements = handleEventMessage(format, storage, line, numberOfElements);
+        }
+        else {
+            numberOfElements = handleTaskMessage(format, storage, line, numberOfElements);
+        }
+        return numberOfElements;
+    }
```
Consider using switch case for less code + better readability

weiliangng commented on src/main/java/LineReader.java:
```diff
@@ -0,0 +1,122 @@
+        String line;
+        Scanner in = new Scanner(System.in);
+        int numberOfElements = 0;
+        while (status) {
```
why are list commands handled differently from say other commands like todo? would be better to standardize everything into 

handleByeCommand
handleEventCommand etc

vincesum replied:
"bye", "list", and "help" are single-word commands that do not need to be handled, hence they are handled differently.

weiliangng commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,39 @@
-| |_| | |_| |   <  __/
-|____/ \__,_|_|\_\___|
-
+    Hello! I'm Vinbot ^-^
```
good use of testing

weiliangng commented on src/main/java/Vinbot.java:
```diff
@@ -0,0 +1,48 @@
+        return new printWelcomeMessage(emptyLine, starLine, spacing);
+    }
+
+    public static void printHelpMessage(printWelcomeMessage format) {
+        System.out.println(format.spacing() + "Enter \"list\" to view your current tasks");
+        System.out.println(format.spacing() + "Enter \"mark\" followed by a number to mark task as complete");
+        System.out.println(format.spacing() + "To enter a task, enter \"todo\" followed by your task to store task");
+        System.out.println(format.spacing() + "To enter a deadline, enter \"deadline\" followed by your task and \"/by\" due date to store deadline");
+        System.out.println(format.spacing() + "To enter an event, enter \"event\" followed by your event and \"/from\" start date \"to\" end date to store event");
+        System.out.print(format.spacing() + "Enter \"unmark\" followed by a number to mark task as incomplete\n");
+        System.out.print(format.spacing() + format.emptyLine() + "\n");
+    }
```
since your handleXcommands also do the printing, can also standardize these for your help command too

weiliangng commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+public class Task {
```
good use of abstraction

weiliangng reviewed (commented):
Good work keep it up!

vincesum committed:
Fix bug for text-ui testing

Update naming conventions based on PR comments

vincesum committed:
Error handling for todo and mark

vincesum committed:
Bug fix for handleMarkMessage

vincesum committed:
Error handling completed for Deadlines and Events. Now program throws VinException whenever an invalid format of String is enterred

vincesum committed:
Merge branch 'branch-Level-5'

vincesum committed:
Package Vinbot and the respective classes

vincesum committed:
Merge branch 'branch-A-Packages'

vincesum committed:
Added indentations for some error messages

TVageesan commented on src/main/java/Vinbot/Tasks/Deadlines.java:
```diff
@@ -0,0 +1,38 @@
+
+import Vinbot.Vinbot;
+
+public class Deadlines extends Task {
```
The name "Deadline" is a more appropriate name for the class. To put it intuitively, the class name "Deadlines" implies one object contains multiple deadlines, when in reality each Deadline object only contains one deadline. This logic can also be applied to your "Events" and "Todos" class names.

TVageesan commented on src/main/java/Vinbot/Tasks/Deadlines.java:
```diff
@@ -0,0 +1,38 @@
+
+    public static String[] scan(Vinbot.printWelcomeMessage format, String line) throws VinException {
+        if (!line.contains("/by")) { //Remove invalid cases
+            throw new VinException(format.spacing() + format.emptyLine() + "\n" + format.spacing() + "invalid deadline not added >.<\n" + format.spacing() + format.starLine() + "\n");
```
I like the initiative to do custom formatting for your error messages! However, you repeat this logic each time you throw a new VinException, with the only difference being the message in the middle. This is a common pattern (in software jargon we call this a [code smell](https://refactoring.guru/smells/duplicate-code)) that tells us there might be room for abstraction here.

Since you've already gone to the effort to make a custom exception, is there a way to wrap this custom formatting logic inside the exception such that you can just call the exception like so?
 `throw new VinException("invalid deadline not added >.<")`

TVageesan commented on src/main/java/Vinbot/Tasks/Deadlines.java:
```diff
@@ -0,0 +1,38 @@
+        if (desc.trim().isEmpty()) {
+            throw new VinException(format.spacing() + format.emptyLine() + "\n" + format.spacing() + "No description entered for deadline!!!\n" + format.spacing() + format.starLine() + "\n");
+        }
+        String date = line.substring(line.indexOf("/by") + 3).trim();
```
The number "3" here is a magic number and the string "/by" is a magic string. To improve readability, perhaps a constant can be used to denote each value.

TVageesan commented on src/main/java/Vinbot/Tasks/LineReader.java:
```diff
@@ -0,0 +1,157 @@
+            }
+            line = in.nextLine();
+            switch (line.toLowerCase()) {
+            case "bye":
```
The command strings "bye", "list" etc. are magic strings. perhaps a Command enum could help readability here.

TVageesan reviewed (commented):
Overall, looks good! I can tell you've put a lot of effort into your string formatting and personalizing your bot, which is great. I believe if you're looking to improve the code even further, one thing you could focus on is abstracting common string formatting functionality into their own methods. You could even consider a dedicated UI class for all printing related functionality.

vincesum committed:
Update exception

vincesum committed:
Change storage for task to ArrayList<Task>

vincesum committed:
Add handleDeleteMessage and implement delete function

vincesum committed:
Add FileEditor

Add makefile, writeToFile and revertToText methods to FileEditor.

revertToText handles todo, deadline and events but not mark/unmark

vincesum committed:
Add scanner to scan Vinbot.txt

Allow Vinbot.txt to save status of marked files
Fix bug where numberOfElements is not updated properly
Add class SystemOutDisabler to disable system output when Vinbot reads Vinbot.txt

vincesum committed:
gitignore update

vincesum committed:
Merge branch 'branch-level-6'

vincesum committed:
Merge branch 'branch-level-7'

# Conflicts:
#	src/main/java/Vinbot/Tasks/LineReader.java
#	src/main/java/Vinbot/Vinbot.java

vincesum committed:
Update gitignore

vincesum committed:
Wrap custom string formatting into Exception

vincesum committed:
Remove magic strings and numbers from Deadline and Event class.

vincesum committed:
Create classes MessageFormat, Storage and Parser to implement more OOP.

Change signature of methods in all classes that use printWelcomeMessage to use MessageFormat class instead

vincesum committed:
Create TaskList class

vincesum committed:
Remove format and numberOfElements in method signature.

format and numberOfElements can be accessed using classes MessageFormat and Tasklist respectively

vincesum committed:
Add UI class and refactored printing of output to be handled solely by UI class

vincesum committed:
Comment on methods of Vinbot, Deadline, Event and Parser

vincesum committed:
Add date class

Fix bug where events and deadlines does not save in storage due to incorrect formatting of storage.

vincesum committed:
Add find method to Parser class

Add handleFindMessage and error handling to find tasks

vincesum committed:
Merge pull request #2 from vincesum/branch-Level-9

Add find method to Parser class

vincesum committed:
Merge pull request #1 from vincesum/branch-Level-8

Add date class

vincesum committed:
Add comments for Parser, Storage, UI, Deadline, Event and Task classes

vincesum committed:
Add comments to adhere to coding standards

vincesum committed:
Update README.md for User Guide for github pages

vincesum committed:
Change formatting for README.md

vincesum committed:
Update README and adjust format

vincesum committed:
Merge branch 'branch-A-JavaDoc'

vincesum committed:
Add image to page

vincesum committed:
Update README.md to improve formatting and printHelpMessage in UI to include delete and find

vincesum committed:
Update Javadoc comments
